tests/e2e/network/ensureloadbalancer.go

```diff
@@ -740,6 +744,315 @@ var _ = Describe("Ensure LoadBalancer", FlakeAttempts(3), Label(utils.TestSuiteL
+		ip := createAndExposeDefaultServiceWithAnnotation(cs, testServiceName, ns.Name, labels, annotation, ports)
+		service, err := cs.CoreV1().Services(ns.Name).Get(context.TODO(), testServiceName, metav1.GetOptions{})
+		defer func() {
+			By("Cleaning up")
+			err := utils.DeleteServiceIfExists(cs, ns.Name, testServiceName)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+
+		By("Retrieving etags from resources")
+		lbEtag, _, _ := getResourceEtags(tc, ip, cloudprovider.DefaultLoadBalancerName(service), true)
+
+		By("Adding a dummy annotation to trigger a second service reconciliation")
+		Expect(err).NotTo(HaveOccurred())
+		annotation["dummy-annotation"] = "dummy4"
+		service = updateServiceAnnotation(service, annotation)
+		utils.Logf("service's annotations: %v", annotation)
+		_, err = cs.CoreV1().Services(ns.Name).Update(context.TODO(), service, metav1.UpdateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		ip, err = utils.WaitServiceExposureAndValidateConnectivity(cs, ns.Name, testServiceName, ip)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Checking etags are not changed")
+		newLbEtag, _, _ := getResourceEtags(tc, ip, cloudprovider.DefaultLoadBalancerName(service), true)
+		Expect(lbEtag).To(Equal(newLbEtag))
+	})
+})
+
+func createNewSubnet(tc *utils.AzureTestClient, subnetName string) (*network.Subnet, bool) {
+	vNet, err := tc.GetClusterVirtualNetwork()
+	Expect(err).NotTo(HaveOccurred())
+
+	var subnetToReturn *network.Subnet
+	isNew := false
+	for i := range *vNet.Subnets {
+		existingSubnet := (*vNet.Subnets)[i]
+		if *existingSubnet.Name == subnetName {
+			By("Test subnet have existed, skip creating")
+			subnetToReturn = &existingSubnet
+			break
+		}
+	}
+
+	if subnetToReturn == nil {
+		By("Test subnet doesn't exist. Creating a new one...")
+		isNew = true
+		newSubnetCIDR, err := utils.GetNextSubnetCIDR(vNet)
+		Expect(err).NotTo(HaveOccurred())
+		newSubnet, err := tc.CreateSubnet(vNet, &subnetName, &newSubnetCIDR, true)
+		Expect(err).NotTo(HaveOccurred())
+		subnetToReturn = &newSubnet
+	}
+
+	return subnetToReturn, isNew
+}
+
+func getResourceEtags(tc *utils.AzureTestClient, ip, nsgRulePrefix string, internal bool) (lbEtag, nsgEtag, pipEtag string) {
+	if internal {
+		lbEtag = to.String(getAzureInternalLoadBalancerFromPrivateIP(tc, ip, "").Etag)
+	} else {
+		lbEtag = to.String(getAzureLoadBalancerFromPIP(tc, ip, tc.GetResourceGroup(), "").Etag)
+	}
+
+	nsgs, err := tc.GetClusterSecurityGroups()
+	Expect(err).NotTo(HaveOccurred())
+	for _, nsg := range nsgs {
+		if nsg.SecurityRules == nil {
+			continue
+		}
+		for _, securityRule := range *nsg.SecurityRules {
+			utils.Logf("Checking security rule %q", to.String(securityRule.Name))
+			if strings.HasPrefix(to.String(securityRule.Name), nsgRulePrefix) {
+				nsgEtag = to.String(nsg.Etag)
+				break
+			}
+		}
+	}
+
+	if !internal {
+		pip, err := tc.GetPublicIPFromAddress(tc.GetResourceGroup(), ip)
+		Expect(err).NotTo(HaveOccurred())
+		pipEtag = to.String(pip.Etag)
```

Comment:
In `GetPublicIPFromAddress`:
`pipList, err := azureTestClient.ListPublicIPs(resourceGroupName)`

Is it possible that `pipList` is empty and returned `pip` is nil?

Reply:
No, in that case, pip would be an empty object (not nil) and pip.Etag would be nil.